Improve ColorPicker picker shape keyboard and joypad accessibility

FeniXb3 opened the pull request:
While experimenting with accessibility, I realized that ColorPicker does not allow you to move the picker cursor around the color rectangle/wheel with keyboard or joypad (and I needed it, as it's the only part of ColorPicker that I want to show).

This PR implements:
- ability of both parts of picker shape to grab focus,
- setting picker focus StyleBox in theme,
- modifying picker cursor position with keyboard or joypad,
- keeping focus on proper control while changing picker shape,
- automatic focusing on the picker shape if it is visible and the HTML LineEdit is not visible (as it's what is currently being focused by default).

How it works in game and in the editor:

https://github.com/user-attachments/assets/0361be2d-364b-44d5-95d3-aa5ed930138a


https://github.com/user-attachments/assets/d5660772-ffb1-48e3-9817-60b75c14941e



I don't really like how it works with the controller, as I'm using `gui_input` to check for actions and the controller does not send echo events, but it's possible to use the controller at least.

FeniXb3 committed:
Modify color in ColorPicker with keyboard or joypad

fire reviewed (approved):
I support this usability improvement in the proposal design, but it needs technical review.

AThousandShips commented on scene/gui/color_picker.cpp:
```diff
@@ -1400,6 +1465,85 @@ void ColorPicker::_uv_input(const Ref<InputEvent> &p_event, Control *c) {
+			color_change_vector.y += 1;
+		}
+
+		if (!Math::is_zero_approx(color_change_vector.length())) {
```
```suggestion
		if (!Math::is_zero_approx(color_change_vector.length_sqared())) {
```
No need for a square root here

FeniXb3 replied:
I realized I could make it even easier and more readable and changed it to `!color_change_vector.is_zero_approx()`

Calinou commented:
> I don't really like how it works with the controller, as I'm using `gui_input` to check for actions and the controller does not send echo events, but it's possible to use the controller at least.

We should make the controller send fake echo events in this case, so that various GUI nodes can handle these directly. However, this should also be done in a way that doesn't affect existing game logic, which is going to be difficult to handle. Perhaps this should only be done for the built-in actions (those starting with `ui_`)?

Calinou reviewed (approved):
Tested locally, it works as expected.

Some usability comments:

- The hue slider should go faster when you adjust it with the keyboard arrows or gamepad buttons. Compare its speed to the color value on the left:

https://github.com/user-attachments/assets/fca1da03-3bab-499d-8ede-584f23a9d7cd

Ideally, the speed at which the hue is adjusted should match the color value on the left, which is probably 3× to 4× faster. This can be done by increasing the increment for each button press on the hue slider.

- When using a color picker with a wheel shape, moving upwards should "slide" along the circle when you've reached the top (same for any other edge):

https://github.com/user-attachments/assets/f85f6c6f-f3e1-4b13-af6c-29bc2adc3b1d

Right now, it stops so you need to manually move to the left then move back up again.

Calinou commented on scene/gui/color_picker.cpp:
```diff
@@ -1986,13 +2164,23 @@ ColorPicker::ColorPicker() {
 	wheel->set_mouse_filter(MOUSE_FILTER_PASS);
 	wheel->connect(SceneStringName(draw), callable_mp(this, &ColorPicker::_hsv_draw).bind(2, wheel));
 
+	// HACK: I cannot draw focus stylebox on the wheel itself, as it's drawing based on shader.
+	wheel_h_focus_display = memnew(Control);
+	wheel_margin->add_child(wheel_h_focus_display);
+	wheel_h_focus_display->set_mouse_filter(MOUSE_FILTER_PASS);
+	wheel_h_focus_display->set_focus_mode(FOCUS_ALL);
+	wheel_h_focus_display->connect(SceneStringName(draw), callable_mp(this, &ColorPicker::_hsv_draw).bind(3, wheel_h_focus_display));
+	wheel_h_focus_display->connect(SceneStringName(gui_input), callable_mp(this, &ColorPicker::_uv_input).bind(wheel_h_focus_display));
```
We should be able to draw a special circle focus stylebox (using a dedicated StyleBoxFlat in the editor theme/default project theme). This can be left for a future PR as it's not essential.

FeniXb3 replied:
I modified the theme to have separate styleboxes - one for rectangle-shaped elements and another for circle-shaped elements:
![obraz](https://github.com/user-attachments/assets/97bc2f63-30af-41a5-bd71-c44553c63d8e)

![obraz](https://github.com/user-attachments/assets/9ca0e886-8f6a-4a52-8e44-e73593092c51)

FeniXb3 commented:
> The hue slider should go faster when you adjust it with the keyboard arrows or gamepad buttons. Compare its speed to the color value on the left:
> 
> 
> color_value_vs_hue_slider.mp4
> 
> Ideally, the speed at which the hue is adjusted should match the color value on the left, which is probably 3× to 4× faster. This can be done by increasing the increment for each button press on the hue slider.

The hue slider goes exactly 3.6× slower than value slider in another shape, so your estimations were pretty correct. ^^ It is by design - value and saturation have values 0-100, but hue has values from range 0-360. Making each button press to increase hue in bigger increments always would miss some values. But I added a multiplier if the event is echo, so after few repetions it speeds up. But again, as for now it does not affect controllers.

What I'm wondering now is if we should have it configurable in the editor, both for games and for ColorPickers in the editor somehow (in editor settings?). What do you think @Calinou ?

> When using a color picker with a wheel shape, moving upwards should "slide" along the circle when you've reached the top (same for any other edge):
> 
> 
> color_value_wheel_slide.mp4
> 
> Right now, it stops so you need to manually move to the left then move back up again.

Fixed, please check it now:

https://github.com/user-attachments/assets/85d85fb5-f762-4709-b27c-e413fa811b07



I thought about making the cursor to spin around on the edge if you keep holding the same button, but while it went well for the HSV Wheel shape, it was buggy for both Circle shapes, so I stashed it for now. This is how it looked for HSV Wheel:

https://github.com/user-attachments/assets/7780bdbe-7e4e-4f8a-8a51-23dc5c7450ed

I am not sure if it is worth to try implementing it well for all Circle shapes or skip it, as I don't know if anyone will take advantage of this feature.

Calinou commented:
> What I'm wondering now is if we should have it configurable in the editor, both for games and for ColorPickers in the editor somehow (in editor settings?). What do you think @Calinou ?

This would be a pretty niche option, so I'd focus on picking a good default instead.

KoBeWi commented:
#82979 could be used for joypad navigation _if it was merged_. For now the only way is implement the behavior using NOTIFICATION_PROCESS manually, like it was done in Slider for example.

Though the joypad navigation is a bit janky right now. When you enter the main color box, the only way out is either accepting or canceling; you can no longer select another element. There could be 2 focus modes here: one for navigating the elements and one for changing colors using the elements. You could switch them with accept/cancel. We have that in LineEdit.
Though it's something that can be added later.

KoBeWi commented:
The wheel is difficult to use.
![image](https://github.com/user-attachments/assets/b2aba832-372b-4560-9766-34c9fccc940b)
You need to press the direction depending on the current rotation. It would be easier if you could press a button to start moving and then it would rotate as long as it's still pressed, without needing to press other buttons.

KoBeWi commented on scene/gui/color_picker.h:
```diff
@@ -267,6 +274,8 @@ class ColorPicker : public VBoxContainer {
 	void _sample_draw();
 	void _hsv_draw(int p_which, Control *c);
 	void _slider_draw(int p_which);
+	int get_wheel_h_change(Vector2 color_change_vector);
```
Unused.

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -1246,11 +1308,11 @@ void ColorPicker::_hsv_draw(int p_which, Control *c) {
 			Vector<Point2> points;
 			Vector<Color> colors;
 			Color col;
-			col.set_ok_hsl(ok_hsl_h, ok_hsl_s, 1);
+			col.set_ok_hsl(h, s, 1);
 			Color col2;
-			col2.set_ok_hsl(ok_hsl_h, ok_hsl_s, 0.5);
+			col2.set_ok_hsl(h, s, 0.5);
 			Color col3;
-			col3.set_ok_hsl(ok_hsl_h, ok_hsl_s, 0);
+			col3.set_ok_hsl(h, s, 0);
```
That doesn't look correct.

FeniXb3 replied:
It must have been merging artifact, it looks like I accidentally missed those changes when resolving a merge conflict from master and forced the old version. Great that you found it, it's fixed now.

KoBeWi commented on scene/gui/color_picker.h:
```diff
@@ -267,6 +274,8 @@ class ColorPicker : public VBoxContainer {
 	void _hsv_draw(int p_which, Control *c);
 	void _slider_draw(int p_which);
+	int get_wheel_h_change(Vector2 color_change_vector);
+	float get_h_on_wheel(Vector2 color_change_vector);
```
```suggestion
	float get_h_on_wheel(const Vector2 &color_change_vector);
```

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -1414,6 +1531,81 @@ void ColorPicker::_uv_input(const Ref<InputEvent> &p_event, Control *c) {
 		}
 	}
+
+	// TODO: Is it better to cast and check it or to not check it and get vector?
```
You need to check whether the arriving even is a pressed action. But afterwards you could probably use `Input.get_vector()` and it might yield better results.

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -1414,6 +1531,81 @@ void ColorPicker::_uv_input(const Ref<InputEvent> &p_event, Control *c) {
+			} else if (actual_shape == SHAPE_VHS_CIRCLE || actual_shape == SHAPE_OKHSL_CIRCLE) {
+				Vector2 center = c->get_size() / 2.0;
+
+				// HACK: It's a hack, as it messes up if I calculate it this way always.
```
What does HACK refer to? `Vector2i()` is initial value of the cursor vector, so it needs to be initialized. Not sure what the other condition is for, it seems to have no effect.

FeniXb3 replied:
At first, I wanted to save the cursor position calculated in `_hsv_draw` and use it in `_uv_input`, but the cursor was moving in strange directions, as I mentioned in another comment. I checked if calculating it here every time would fix the issue, but it didn't. That's why I decided to calculate it only if it was not set (or was reset by clicking with mouse) and this is what the hack refers to, as it  is not the way I think it should be done, but it does not seem to work well with the correct way. I might 

> Not sure what the other condition is for, it can be safely removed it seems.
I added it when I was working on sliding along the edge of the circle when you keep pressing any direction. But it seems it is not needed now. Removed.

KoBeWi commented on scene/gui/color_picker.h:
```diff
@@ -117,6 +117,10 @@ class ColorPicker : public VBoxContainer {
 #endif
 
 	int current_slider_count = SLIDER_COUNT;
+	// TODO: Think about better name or a way to not use it at all
+	Vector2i hsv_keyboard_picker_cursor_position;
```
A way to not use it is keeping the variable local in `_uv_input()`, which means calculating sin/cos again every time.

FeniXb3 replied:
The hack mentioned in another place is connected with issues with calculating sin/cos again every time. I tried it at the beginning and it's behaviour was strange - the cursor was not going in straight lines. I believe it might be caused by some precision loss while calculating this, or something like that.

I realized that even when you drag the cursor with mouse, the offset of it's center and the mouse cursor is changing, depending on how high or low on the circle you are. You can see that on this video:

https://github.com/user-attachments/assets/baade90a-bea1-48d4-bff2-f3b151e62b7f

FeniXb3 commented:
> #82979 could be used for better joypad navigation _if it was merged_. For now the only way is implement the behavior using NOTIFICATION_PROCESS manually, like it was done in Slider for example.

I did it like in Slider, thanks for the suggestion. I tried with `get_vector` in `NOTIFICATION_INTERNAL_PROCESS` before, but it was not working. Now I see that I might have missed calling `set_process_internal(true)` in the right moment.

> Though the joypad navigation is a bit janky right now. When you enter the main color box, the only way out is either accepting or canceling; you can no longer select another element. There could be 2 focus modes here: one for navigating the elements and one for changing colors using the elements. You could switch them with accept/cancel. We have that in LineEdit. Though it's something that can be added later.

I added this second focus mode like in LineEdit. I thought about it before, but I didn't see benefits in my usecase so I skipped it before. It works well, for both LineEdit and ColorPicker picker shapes now if you use a ColorPicker directly. But if you show it via ColorPickerButton, it failes in both cases. The reason is that both accept and cancel actions close the popup that ColorPicker is in. You can see the behaviour on this video:

https://github.com/user-attachments/assets/7eeff182-4fee-495d-96ca-35dde341a4c3



I didn't bother to implement this as I use focus next/prev actions to navigate through ColorPicker's controls. I configured inputs for those actions to shoulder buttons and I was using them to test this feature. I forgot that by default there are no joypad buttons connected to those actions. Is there any reason this is not done by default? I wonder if I can simply configure them within this PR to make it work better out of the box.



> The wheel is difficult to use. ![image](https://private-user-images.githubusercontent.com/2223172/391957458-b2aba832-372b-4560-9766-34c9fccc940b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzM2MDYwMzAsIm5iZiI6MTczMzYwNTczMCwicGF0aCI6Ii8yMjIzMTcyLzM5MTk1NzQ1OC1iMmFiYTgzMi0zNzJiLTQ1NjAtOTc2Ni0zNGM5ZmNjYzk0MGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTIwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEyMDdUMjEwODUwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTU1N2MzMjA1YWU4ZGQ2YjMzZGE5YWVkZTNhY2FhMzZiNmFjNjM4NWE5MDA3NjI2ZDI2NjM1ODRjOGE1ZDI4NiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.kK-S_CcwQAToJ_jg7Bi8J-PhMxFXh4C4pKL7n5BMkDc) You need to press the direction depending on the current rotation. It would be easier if you could press a button to start moving and then it would rotate as long as it's still pressed, without needing to press other buttons.

Done. I had it stashed, as I mentioned in one of the previous comments here, as I wasn't sure if I should try and implement similar behaviour for both circle shapes. In the end I separated their behaviours.

KoBeWi commented:
> The reason is that both accept and cancel actions close the popup that ColorPicker is in.

You could try using `accept_event()`.

FeniXb3 commented:
> > The reason is that both accept and cancel actions close the popup that ColorPicker is in.
> 
> You could try using `accept_event()` in the shape's gui input, to prevent the event from propagating.

I do. What I see now while debugging is that the `ColorPickerPopupPanel::_input_from_window` gets the input event before my `gui_input` handler so the `accept_event()` don't do much in this case. As tihs popup has no "OK" button, I can understand that accept action is a reasonable way to confirm the color choice,

FeniXb3 commented on scene/gui/color_picker.cpp:
```diff
@@ -1986,13 +2213,23 @@ ColorPicker::ColorPicker() {
 	wheel->set_mouse_filter(MOUSE_FILTER_PASS);
 	wheel->connect(SceneStringName(draw), callable_mp(this, &ColorPicker::_hsv_draw).bind(2, wheel));
 
+	// HACK: I cannot draw focus stylebox on the wheel itself, as it's drawing based on shader.
+	wheel_h_focus_display = memnew(Control);
```
In one moment I did. But it meant that I had to add logic for navigating between two focus modes for this control, capturing next/prev actions, because wheel_uv would be used to handle focus for both parts of HSV Wheel.

It added more complexity to the class that is already overloaded with a lot of ifs and was not working well always (like it was sometimes looping focus between those two states when holding `ui_focus_prev` action instead of leaving the control and switching to the hex LineEdit. I thought that adding another control for both drawing the focus stylebox and making switching focus more reliable is a better choice.

If it's better to have few more ifs than using this empty control, I can work on it again.

KoBeWi replied:
#99515 makes the code cleaner.
If your PR happens to be merged first, I'll simplify it and integrate with my changes.

FeniXb3 replied:
I changed the code to not use a dedicated new Control for handling focus of wheel in HSV Wheel shape. Now the `wheel` control can grab focus in this shape, while focus stylebox is drawn on `wheel_uv`. I had whith `wheel_uv` handling everything and working well with `ui_focus_prev` and `ui_focus_next`, but it `ui_up` was glitching and I choose approach that seemed to work always with some tweaks, without introducing a new control.

FeniXb3 commented:
I believe the feature is done with suggestions applied, if I didn't miss something.

I attach a sample project to make it easier to test it. It has configured ui next and prev actions on controller shoulder buttons and accept/cancel on A/B buttons of Xbox controller.
[godot-colorpicker-accessibility-test.zip](https://github.com/user-attachments/files/18306303/godot-colorpicker-accessibility-test.zip)

AThousandShips commented on scene/gui/color_picker.cpp:
```diff
@@ -153,8 +159,40 @@ void ColorPicker::_notification(int p_what) {
+				gamepad_event_delay_ms -= get_process_delta_time();
+				if (gamepad_event_delay_ms <= 0) {
+					gamepad_event_delay_ms = GAMEPAD_EVENT_REPEAT_RATE_MS + gamepad_event_delay_ms;
+					// Treat any input from joypad axis as -1, 0, or 1, as the value is added to Vector2i and would be lost
```
```suggestion
					// Treat any input from joypad axis as -1, 0, or 1, as the value is added to Vector2i and would be lost.
```

FeniXb3 replied:
Thanks for all your suggestions! I updated this right away and I see those changes [here](https://github.com/godotengine/godot/compare/c90efdb83d9a321aba97271dc0893ff0005d08e8..80de4db26529d5f8712bd268f2f15c6292e2d80d), but it looks like GitHub didn't update this everywhere. It should show all your suggestions as outdated but it doesn't, I don't know why.

AThousandShips replied:
Weird, might be because of your style fixes changing code indentation

FeniXb3 replied:
Ok, everything looks fine now. I realized that others were marked as outdated after few minutes and this one was my ommission, because I have the same comment in two places and I missed one. It's fixed now.

Calinou reviewed (approved):
Tested locally, it works as expected.

Some more feedback from things I noticed. These aren't critical and can be implemented in future PRs, as this PR already provides a decent baseline for keyboard/gamepad usability:

- It doesn't appear to be possible to select these buttons using the keyboard or gamepad only (highlighted in yellow). This includes using <kbd>Tab</kbd> on a keyboard:

![image](https://github.com/user-attachments/assets/43b0a176-0e91-4918-a0d5-a2d961f69c7e)

- Is it possible to ensure the cursor on the color picker always appears in front of the focus outline? Right now, it'll appear behind the focus outline when at the edges of a circular (or square) picker shape.

![image](https://github.com/user-attachments/assets/24d16477-852d-49b9-9b42-07f7845ff818)

KoBeWi commented:
The interactions are much better since I last checked.
I agree that the cursor should appear in front of the focus outline. Other than that the behavior is fine.

KoBeWi commented:
Shape menu can't be opened with gamepad. It appears for 1 frame and disappears. Using keyboard works fine.

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -2615,6 +3031,13 @@ void ColorPresetButton::_notification(int p_what) {
 				WARN_PRINT("Unsupported StyleBox used for ColorPresetButton. Use StyleBoxFlat or StyleBoxTexture instead.");
 			}
+
+			if (this->has_focus()) {
```
```suggestion
			if (has_focus()) {
```

KoBeWi commented on scene/gui/color_picker.cpp:
```diff
@@ -2615,6 +3031,13 @@ void ColorPresetButton::_notification(int p_what) {
 				WARN_PRINT("Unsupported StyleBox used for ColorPresetButton. Use StyleBoxFlat or StyleBoxTexture instead.");
 			}
+
+			if (this->has_focus()) {
+				RID ci = this->get_canvas_item();
```
```suggestion
			if (has_focus()) {
				RID ci = get_canvas_item();
```

FeniXb3 commented:
> Shape menu can't be opened with gamepad. It appears for 1 frame and disappears. Using keyboard works fine. Might be unrelated though.
> 
> EDIT: I think the menu option is accepted immediately for some reason. Same happens with color mode.

That's interesting. Could you send me the project that you use to test it? You can current version of my test projects attached.
[colorpicker_test.zip](https://github.com/user-attachments/files/18630504/colorpicker_test.zip)
I test it on both `ColorPickerButton` and `ColorPicker` straight in the tree, without the popup. The only usability issue I have is that `ui_accept` and `ui_cancel` both close the popup opened by `ColorPickerButton`. It makes most controls in any color picker created by `ColorPickerButton` unusable. You cannot change the shape, add preset or load palette, because it would close the popup right away.
I wonder if we can introduce another action for closing the popup that would be bound to Ctrl+Enter or something like that. What do you think?

KoBeWi commented:
I use a random project with default ColorPicker on the scene. It's not project-specific.

> I wonder if we can introduce another action for closing the popup that would be bound to Ctrl+Enter or something like that. What do you think?

I think the main problem is that `ui_accept` will close the popup even if accepting something else, e.g. when you edit SpinBox value and press Enter, it will both accept the value and close the popup. The picker popup should be aware of focus and close only when not accepting input for another control. That's for another PR though.

FeniXb3 commented:
All controls mentioned by @Calinou should be accessible with keyboard and joypad now. You can now also add and remove presets. I added a new action for deleting preset, `ui_colorpicker_delete_preset` with default events `Key::KEY_DELETE` and `JoyButton::X`.

Tooltip for preset button describes how to use or remove the color with mouse:
![Zrzut ekranu 2025-02-01 084026](https://github.com/user-attachments/assets/f7bb9239-e8e1-4387-aace-4cd3dfc14203)

I have not pushed changes that dynamically add information about other events connected with deleting the color preset:
![Uploading Zrzut ekranu 2025-02-01 205542.png…]()
I am not sure if this is a good idea. And if yes, should I manually modify translations in all *.po files to include extra `%s` or is there another workflow for that?

Additionally, I finally decided to add default joypad events for:
- `ui_cancel` - `JoyButton::B`
- `ui_accept` - `JoyButton::A`
- `ui_focus_next` - `JoyButton::RIGHT_SHOULDER`
- `ui_focus_prev` - `JoyButton::LEFT_SHOULDER`

KoBeWi commented:
> And if yes, should I manually modify translations in all *.po files to include extra %s or is there another workflow for that?

Translations are edited on Weblate. They are synchronized by maintainers, you should not edit .po files manually.

FeniXb3 commented:
> Shape menu can't be opened with gamepad. It appears for 1 frame and disappears. Using keyboard works fine. Might be unrelated though.
> 
> EDIT: I think the menu option is accepted immediately for some reason. Same happens with color mode.

> I use a random project with default ColorPicker on the scene. It's not project-specific.
> 
> EDIT: Ok you need to disable `display/window/subwindows/embed_subwindows` project setting for the bug to happen.
> 
I realized that this is due to the nature of gamepad input being sent to every window whether it's focused or not. With this option disabled each popup or popup menu are treated as separate windows and all handle gamepad input, so `ui_acccept` is being processed by both the main window and the popup kinda simultaneously. You can see this behaviour on the video below:

https://github.com/user-attachments/assets/3d487215-17e0-4936-9ef6-315e07ec3fe1

It means that any UI with popups is inaccessible with gamepad if "embed subwindows" option is disabled. I will see if I can do some workarounds for `ColorPicker`, but it looks like a more general input design change for not embeded popups for a separate discussion and PR.

FeniXb3 commented:
I rebased to the newest master and it seems that issues with `PopupMenu`'s popup capturing `ui_accept` event from gamepad that opened it were fixed by someone else in the meantime.
I pushed my changes that fix another issue when subwindows are not embedded that I showed before. Input on the main window is now blocked when any `PopupMenu` is opened in `ColorPicker`. I believe it would be good to handle it more generally in seperate PR.

I started to work on allowing picking a color from the screen with keyboard or gamepad, but it is not included here yet, as I have some cases still to handle.

https://github.com/user-attachments/assets/315b843c-29d1-4c5d-82b2-4a943d3782fd


Will it be better to have a separate PR for this or should I keep on working on this here?

KoBeWi commented:
This PR is already approved, so it's better to leave further changes for later if they are not necessary.

FeniXb3 commented:
@Repiteo thanks for the info, I missed it. It's rebased now.

Repiteo commented:
Thanks! Congratulations on your first merged contribution! 🎉
